Add optimization of numpy code involving matrix inverses

anpandey opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### Brief description of what is fixed or changed
Adds an `optimization` option to the code printer base class that executes any `codegen.rewriting` optimizations the printer object might have. Also adds an optimization to the numpy printer that uses the assumption system to rewrite the multiplication of a matrix and a vector as the faster `np.linalg.solve` when appropriate.

For example:

```python3
>>> from sympy.assumptions import assuming, Q
>>> from sympy.matrices import MatrixSymbol
>>> from sympy.printing.pycode import NumPyPrinter
>>> A = MatrixSymbol('A', 2, 2)
>>> x = MatrixSymbol('x', 2, 1)
>>> p = NumPyPrinter(settings={'optimize' : True})
>>> print(p.doprint(A**(-1) * x))
(numpy.linalg.inv(A)).dot(x)
>>> with assuming(Q.fullrank(A)):
...     p.doprint(A**(-1) * x)
numpy.linalg.solve(A, x)
```

#### Other comments

This PR is mostly meant as an exploration of how (and whether) the `codegen.rewriting` module is enough for generating optimizations. I'd appreciate comments and suggestions about this approach.

#### Release Notes


<!-- BEGIN RELEASE NOTES -->
* printing
  * Add support for optimizations to numpy printer.
  * The numpy and octave printers print matrix expressions like `A**-1*x` using a `solve`.
<!-- END RELEASE NOTES -->

anpandey committed:
Add optimization of numpy code involving matrix inverses

Adds an `optimization` option to the code printer base class that
executes any `codegen.rewriting` optimizations the printer object might
have. Also adds an optimization to the numpy printer that uses the
assumption system to rewrite the multiplication of a matrix and a vector
as the faster `np.linalg.solve` when appropriate.

For example:

    >>> from sympy.assumptions import assuming, Q
    >>> from sympy.matrices import MatrixSymbol
    >>> from sympy.printing.pycode import NumPyPrinter
    >>> A = MatrixSymbol('A', 2, 2)
    >>> x = MatrixSymbol('x', 2, 1)
    >>> p = NumPyPrinter(settings={'optimize' : True})
    >>> print(p.doprint(A**(-1) * x))
    (numpy.linalg.inv(A)).dot(x)
    >>> with assuming(Q.fullrank(A)):
    ...     p.doprint(A**(-1) * x)
    numpy.linalg.solve(A, x)

asmeurer commented:
Last I saw Travis didn't run on draft pull requests, but maybe that has been fixed because it looks like it ran here. Anyway, if you have any issues where Travis doesn't run, you may need to make the PR not a draft (just use `WIP` in the title instead).

asmeurer commented on sympy/printing/pycode.py:
```diff
@@ -488,6 +490,16 @@ class NumPyPrinter(PythonCodePrinter):
+        # XXX: Needed to avoid circular import
+        from sympy.codegen.rewriting import ReplaceOptim
+
+        self.optimizations = [
```
We usually put these things on the class rather than the instance. Although maybe that's not such a great thing, because putting a mutable data structure on a class can be problematic if a subclass mutates it.

asmeurer commented:
@bjodah any thoughts on this approach?

bjodah commented:
Definitely an interesting approach. I only had time to glance over it at this time. Are the optimizations exclusive to numpy? Wouldn't the octave printer benefit from similar rewrites (I don't know if it applies to R/julia/mathematica too)? In that case the predicate/transformers, which can be shared, could go somewhere else under `sympy.codegen` possibly `sympy.codegen.rewriting`.

Then there's the question of whether the logic should be a printer option or something that the user applies themselves (separation of concerns, and keeping the complexity of the printers at bay). I'm leaning toward the latter, at least that's what we [ended up doing](https://github.com/sympy/sympy/pull/14139#issuecomment-366317061) for `create_expand_optimization`. It is always simpler adding new helpers which apply different transformations (if the learing curve is too steep) at a later stage, than it is to revert APIs which allow for everything but the kitchen sink.

Either way optimize=True/False will likely not be fine-grained enough?

anpandey commented:
@bjodah Right now, the optimizations are exclusive to numpy. The optimization would also probably work with the Octave printer, but Octave accomplishes this with an arithmetic operator (`\`) instead of a function call, which seems to be unsupported by the Octave printer for now. I'm not too familiar with any of the other languages, so I'm focusing on numpy for now.

My motivation for combining the code printer and optimizations was because each language might have its own way of representing an operation (such as with Octave) and a given optimization might apply to only a certain language. Now that I think about it, we could probably abstract this away with a specialized AST node (like `MatSolve`) and printers for each language. This would probably also solve the issue of the fine-grainedness of `optimize={True,False}`.

bjodah commented:
I'm no expert in those languages either, maybe someone else can chime in there. A `MatSolve` node sounds like a good idea to me.

anpandey committed:
Move to representing solve as an AST node

Removes the `optimize` option introduced to the code printer in the
previous commit, instead instroducing the inverse optimization as a
top-level variable in `codegen.rewriting` that has to be applied by the
user. Also abstracts the solving operation as a `MatrixSolve` AST node
that can be handled by each code printer uniquely.

anpandey commented:
The build failed because of a circular import -- the `MatrixSolve` node needs to subclass `MatrixExpr` for matrix expressions involving it to be considered valid. I'm not exactly sure how I can approach this.

bjodah commented:
Instead of putting it in ast.py you can put it in another file which isn't imported early. We have C-specific nodes in cnodes.py, Fortran specific nodes in fnodes.py and C++ specific ones in cxxnodes.py

This might span multiple languages so perhaps a "common_nodes.py" or "matrix_operation_nodes.py" file?

anpandey committed:
Move MatrixSolve to its own file and fix tests

anpandey committed:
Add arg test for MatrixSolve in test_args

anpandey committed:
Add `MatrixSolve` printing support to Octave printer

anpandey committed:
Fix wrong import in test

anpandey committed:
Add some documentation

bjodah commented:
Great work! I really like the approach in the last couple of commits, looking forward to seeing more of these kinds of transformations.

oscarbenjamin commented:
@bjodah is this good to merge?

bjodah commented:
The title has [WIP] in it? Other than that I think it looks good.

anpandey commented:
@bjodah I think it's ok to be merged now -- I'll address some other issues in a different PR.

bjodah commented:
Sounds good, looks like there's a merge conflict so maybe merge master into your branch?

anpandey committed:
Merge branch 'master' of https://github.com/sympy/sympy into numpy_optim

asmeurer commented:
I've updated the release notes. This looks good to me.
